File: code/backend/python/app/adapters/rest/match/experience_controller.py
"""Step 38 — POST /api/gameplay/{uuid_match}/action/use-exp (mirror of the Java ExperienceController)."""
import time
from typing import Any, Dict

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from app.core.ports.match.experience_ports import ExperienceError, ExperiencePort

_STATUS_BY_CODE = {
    ExperienceError.MATCH_NOT_FOUND: 404,
    ExperienceError.INVALID_STAT: 400,
    ExperienceError.MATCH_NOT_RUNNING: 409,
    ExperienceError.NOT_YOUR_TURN: 409,
    ExperienceError.COMA: 409,
    ExperienceError.SLEEPING: 409,
    ExperienceError.LOCATION_NOT_SAFE: 409,
    ExperienceError.MAX_STAT_VALUE: 409,
    ExperienceError.NOT_ENOUGH_EXP: 409,
}


def _error(code: str, message: str, http_status: int) -> JSONResponse:
    return JSONResponse(
        status_code=http_status,
        content={"error": code, "message": message, "timestamp": int(time.time() * 1000)},
    )


def _result_to_camel(r: Dict[str, Any]) -> Dict[str, Any]:
    return {
        "matchUuid": r.get("match_uuid"),
        "characterUuid": r.get("character_uuid"),
        "stat": r.get("stat"),
        "statBefore": r.get("stat_before"),
        "statAfter": r.get("stat_after"),
        "expBefore": r.get("exp_before"),
        "expAfter": r.get("exp_after"),
        "expCost": r.get("exp_cost"),
        "expCosts": dict(r.get("exp_costs") or {}),
        "statChanges": [
            {"characterUuid": c.get("character_uuid"), "statistic": c.get("statistic"),
             "before": c.get("before"), "after": c.get("after"), "delta": c.get("delta")}
            for c in (r.get("stat_changes") or [])
        ],
    }


class ExperienceController:
    def __init__(self, experience_port: ExperiencePort):
        self.experience_port = experience_port
        self.router = APIRouter()
        self.router.add_api_route(
            "/api/gameplay/{uuid_match}/action/use-exp", self.use_exp, methods=["POST"])

    async def use_exp(self, uuid_match: str, request: Request):
        user_uuid = getattr(request.state, "user_uuid", None)
        if not user_uuid:
            return _error("UNAUTHENTICATED", "User identity is missing", 401)
        try:
            body = await request.json()
        except Exception:
            body = {}
        stat = (body or {}).get("stat") if isinstance(body, dict) else None
        if not stat or not str(stat).strip():
            return _error("INVALID_STAT", "stat is required: one of dex, int, cos", 400)
        try:
            result = self.experience_port.use_exp(uuid_match, user_uuid, str(stat))
        except ExperienceError as exc:
            return _error(exc.code, exc.message, _STATUS_BY_CODE.get(exc.code, 409))
        return JSONResponse(status_code=200, content=_result_to_camel(result))

File: code/backend/java/adapter-rest/src/main/java/games/paths/adapters/rest/controller/match/ExperienceController.java
package games.paths.adapters.rest.controller.match;

import games.paths.adapters.rest.dto.UseExpRequest;
import games.paths.adapters.rest.dto.UseExpResponse;
import games.paths.core.port.match.ExperiencePort;
import games.paths.core.port.match.ExperiencePort.ExperienceException;

import jakarta.servlet.http.HttpServletRequest;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * ExperienceController - REST adapter of Step 38: POST /api/gameplay/{uuidMatch}/action/use-exp
 * spends experience on a +1 of dex / int / cos. Body {@code {"stat": "dex"}}.
 */
@RestController
public class ExperienceController {

    private final ExperiencePort experiencePort;

    public ExperienceController(ExperiencePort experiencePort) {
        this.experiencePort = experiencePort;
    }

    @PostMapping("/api/gameplay/{uuidMatch}/action/use-exp")
    public ResponseEntity<Object> useExp(@PathVariable String uuidMatch,
                                         @RequestBody(required = false) UseExpRequest body,
                                         HttpServletRequest request) {
        String userUuid = userUuid(request);
        if (userUuid == null) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED", "User identity is missing");
        }
        String stat = body != null ? body.getStat() : null;
        if (stat == null || stat.isBlank()) {
            return error(HttpStatus.BAD_REQUEST, "INVALID_STAT", "stat is required: one of dex, int, cos");
        }
        try {
            return ResponseEntity.ok(UseExpResponse.fromModel(experiencePort.useExp(uuidMatch, userUuid, stat)));
        } catch (ExperienceException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    private static String userUuid(HttpServletRequest request) {
        String u = (String) request.getAttribute("userUuid");
        return (u == null || u.isBlank()) ? null : u;
    }

    private static ResponseEntity<Object> error(HttpStatus status, String code, String message) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("error", code);
        body.put("message", message);
        body.put("timestamp", System.currentTimeMillis());
        return ResponseEntity.status(status).body(body);
    }

    /** Not-found for a missing entity, bad-request for a wrong body, conflict for a state the player could act on. */
    static HttpStatus mapStatus(ExperienceException.Code code) {
        return switch (code) {
            case MATCH_NOT_FOUND -> HttpStatus.NOT_FOUND;
            case INVALID_STAT -> HttpStatus.BAD_REQUEST;
            case MATCH_NOT_RUNNING, NOT_YOUR_TURN, COMA, SLEEPING, LOCATION_NOT_SAFE,
                 MAX_STAT_VALUE, NOT_ENOUGH_EXP -> HttpStatus.CONFLICT;
        };
    }
}
